File: android/sail-android/src/com/appspresso/core/runtime/server/kraken/JsonRpcAsyncPluginContext.java
package com.appspresso.core.runtime.server.kraken;

import java.util.Map;
import org.apache.commons.logging.Log;

import com.appspresso.api.AxLog;
import com.appspresso.api.AxErrorHandler;
import com.appspresso.api.AxRuntimeContext;
import com.appspresso.core.runtime.server.kraken.BridgeSessionManager.BridgeSession;

public class JsonRpcAsyncPluginContext extends JsonRpcPluginContext {

    private final static Log L = AxLog.getLog("JsonRpcAsyncPluginContext");

    private static final String BRIDGE_JSONRPC_METHOD = "ax.bridge.jsonrpc";
    private AxRuntimeContext runtimeContext;

    public JsonRpcAsyncPluginContext(Map<?, ?> requestJson, BridgeSession session) {
        super(requestJson, session);
    }

    public void setRuntimeContext(AxRuntimeContext runtimeContext) {
        this.runtimeContext = runtimeContext;
    }

    @Override
    public void sendResult(Object object) {
        super.makeSuccessResult(object);
        sendRpcResponse(super.getResult());
    }

    @Override
    public void sendError(int code, String message) {
        super.makeErrorResult(code, message);
        sendRpcResponse(super.getResult());
    }

    protected void sendRpcResponse(final Object obj) {
        final BridgeSession session = getSession();
        if (!session.getJavaScriptEvaluationEnabled()) {
            JsonRpcPollHandler.instance().push(session.getSessionID(), obj);
            return;
        }

        AxErrorHandler onError = new AxErrorHandler() {
            @Override
            public void onError(Exception e) {
                if (L.isDebugEnabled()) {
                    L.debug("exception occured while evaluating javascript by loadUrl", e);
                }
                // fall back to long-poll
                session.setJavaScriptEvaluationEnabled(false);
                JsonRpcPollHandler.instance().push(session.getSessionID(), obj);
            }
        };
        runtimeContext.invokeJavaScriptFunction(onError, BRIDGE_JSONRPC_METHOD, obj);
    }

}

File: ios/sail-ios/JsonRpcAsyncPluginContext.m
//
//  JsonRpcAsyncPluginContext.m
//  sail-ios
//
//  Copyright (c) 2012 KTH Corp.
//

#import "JsonRpcAsyncPluginContext.h"
#import "AxRuntimeContext.h"
#import "BridgeSessionManager.h"
#import "RpcPollResultStore.h"
#import "JSONKit.h"

#define BRIDGE_JSONRPC_METHOD   @"ax.bridge.jsonrpc"

@implementation JsonRpcAsyncPluginContext

- (id)initWithRequestJson:(NSDictionary*)json session:(BridgeSession*)session {
    return self = [super initWithRequestJson:json session:session];
}

- (id)setRuntimeContext:(id<AxRuntimeContext>)runtimeContext {
    _runtimeContext = runtimeContext;
    return self;
}

#pragma mark -

- (void)sendRpcResponse:(id)object {
    if (self.session.javaScriptEvaluationEnabled) {
        NSString* js = [NSString stringWithFormat:@"window.setTimeout(function(){%@(%@);},1);", BRIDGE_JSONRPC_METHOD, [object JSONString]];
        [_runtimeContext executeJavaScript:js];
        return;
    }

    [[RpcPollResultStore instance] putResult:object toQueue:self.session.key];
}

- (void)sendResult:(id)result {
    [super makeSuccessResult:result];
    [self sendRpcResponse:self.result];
}

- (void)sendError:(NSInteger)code message:(NSString *)message {
    [super makeErrorResult:code message:message];
    [self sendRpcResponse:self.result];
}

@end
